lib/ukstore/include/uk/store.h

```diff
@@ -29,3 +29,404 @@
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_s8(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_s8_func_t get, uk_store_set_s8_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_u16(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_u16_func_t get, uk_store_set_u16_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_s16(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_s16_func_t get, uk_store_set_s16_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_u32(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_u32_func_t get, uk_store_set_u32_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_s32(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_s32_func_t get, uk_store_set_s32_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_u64(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_u64_func_t get, uk_store_set_u64_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_s64(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_s64_func_t get, uk_store_set_s64_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_uptr(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_uptr_func_t get, uk_store_set_uptr_func_t set);
+struct uk_store_entry *
+		_uk_store_dynamic_entry_create_charp(
+		struct uk_store_folder *folder, const char *name,
+		uk_store_get_charp_func_t get, uk_store_set_charp_func_t set);
+
+/**
+ * Creates a new dynamic entry
+ *
+ * @param folder the folder where to add the entry
+ * @param name the name of the new entry
+ * @param type the type of the new entry
+ * @param get the getter function to save
+ * @param set the setter function to save
+ * @return pointer to the created entry
+ */
+#define uk_store_dynamic_entry_create(folder, name, type, get, set)	\
+	_uk_store_dynamic_entry_create_ ## type((folder), (name), (get), (set))
+
+/**
+ * Creates a new dynamic folder
+ *
+ * @param name the name of the new folder
+ * @return pointer to the created folder
+ */
+#define uk_store_dynamic_folder_create(name)	\
+	_uk_store_dynamic_folder_create((name))
+
+struct uk_store_folder *
+_uk_store_dynamic_folder_create(const char *name);
+
+/**
+ * Adds a dynamic folder to a library
+ *
+ * @param folder the folder to add it to
+ * @return 0 on success and < 0 on failure
+ */
+#define uk_store_add_folder(folder)	\
+	_uk_store_add_folder(uk_store_libid_self(), folder)
+
+int
+_uk_store_add_folder(unsigned int library_id, struct uk_store_folder *folder);
+
+/**
+ * Removes the dynamic folder from the library and frees its memory and
+ * the memory of the entries.
+ *
+ * @param folder the folder to free
+ */
+#define uk_store_free_folder(folder)	\
+	_uk_store_free_folder((folder))
+
+void
+_uk_store_free_folder(struct uk_store_folder **folder);
```

Comment:
No refcounting for folders. I kept the double pointer to mark it as `NULL` to ensure no uses after free.

Reply:
Also, it will be a bit harder to split the `free` into two completely separate parts (`remove`/`free`) as entries that are removed will be lost (for freeing) if they are not freed in the same iteration.

Reply:
For this I will move the `uk_list_del` of the folder inside the `remove` and keep the rest inside the `free`. Does this sound good?